packages/truffle-decode-utils/src/types/values.ts

```diff
@@ -0,0 +1,1423 @@
+import debugModule from "debug";
+const debug = debugModule("decode-utils:values");
+
+//objects for Solidity values
+
+//Note: This is NOT intended to represent every possible value that exists
+//in Solidity!  Only possible values of variables.  (Though there may be
+//some expansion in the future.)  We do however count the builtin variables
+//msg, block, and tx as variables (not other builtins though for now) so
+//there is some support for the magic type.
+
+//We don't include fixed and ufixed for now.  Those will be added when
+//implemented.
+
+//NOTE: not all of these optional fields are actually implemented. Some are
+//just intended for the future.  More optional fields may be added in the
+//future.
+
+//Note: Many of the errors defined here deliberately *don't* extend Error.
+//This is because their intended use is a little different.  Only the ones
+//that are for throwing extend Error.
+
+import BN from "bn.js";
+import { Types } from "./types";
+import util from "util";
+import { AstDefinition } from "../ast";
+import { Definition as DefinitionUtils } from "../definition";
+
+export namespace Values {
```

Comment:
I don't know that anything *forces* them to be in the same namespace, but IMO they should be.  With a possible exception for the underlying errors (*not* the `ErrorResult`s).  They all fundamentally go together, to my mind, just being the different types of `Result` one might get.  Separating them seems more confusing than helpful to me.

Reply:
How bout subnamespaces? I understand the natural grouping you describe, but I'm looking out for lexicographic sorting and figure these could be named big-endianly somehow.

Reply:
I'd say they *are* named big-endianly!  I think of the data type as primary and whether it's a valid value or an error as secondary.  Like, `BoolResult` is a *type*; there's plenty of times you'll get (what you know in advance to be) a `BoolResult` and you don't know beforehand whether it'll be a `BoolValue` or a `BoolErrorResult`, but you definitely know it's a `BoolResult`.  It never happens that you know in advance you'll get a valid value, or know in advance you'll get an error, but don't know the data type.  So `Value` and `ErrorResult` aren't really types in the same sense, even if they're types in the language.  (And if I'm redoing this stuff with union types, I might just discard them entirely.)  So, like I said, I'd say they *are* named in a big-endian fashion.  Data type is primary.

Reply:
I'll withhold further judgment here once you have this file converted over to using `type` aliases. It might look better.

My thinking is just that this is kind of difficult to make sense of:

<details>
<summary>Bag of classes</summary>
<img width="836" alt="Screen Shot 2019-07-01 at 2 58 46 PM" src="https://user-images.githubusercontent.com/151065/60460097-d6488f00-9c10-11e9-8fa8-128c46f33992.png">
My first thought here was "I want to see all the kinds of results, all the kinds of values, etc."
</details>

Reply:
Agree probably best to table this until then.

But I still think you're thinking of this wrong if you want to see "all the kinds of results, all the kinds of values, etc".  Like, the types here that truly act as *types* ares the `Result` classes.  Those should be the primary focus!

...funny thing is, if I redo all this with union types, TypeDoc will put *less* focus on those, considering those to be mere type aliases.  Honestly, I'm not sure I like the idea of structuring my code around what an autodocumentation tool will make of it.

...although this does seem to lean towards the idea that the errors should be separated out to avoid clutter.  I guess if I do that, then that will leave the focus in `value.ts` on the `Value` types, so it could end up more readable after all.

Reply:
I'm going to mark this resolved for now to get it out of the way.